File: cross-platform/windows-controller/src-tauri/src/gradle.rs
use crate::BuildInfo;
use tauri::{AppHandle, Emitter};
use tokio::io::{AsyncBufReadExt, BufReader};
use tokio::sync::broadcast;
use uuid::Uuid;

/// Start a Gradle build, streaming output as Tauri events.
pub async fn start_build(
    project_path: &str,
    task: &str,
    java_home: &str,
    extra_args: Option<&[String]>,
    event_tx: broadcast::Sender<String>,
    app: AppHandle,
) -> Result<BuildInfo, String> {
    let build_id = Uuid::new_v4().to_string();
    let started_at = chrono::Utc::now().to_rfc3339();

    // Determine gradlew path
    let gradlew = if cfg!(windows) {
        format!("{}\\gradlew.bat", project_path)
    } else {
        format!("{}/gradlew", project_path)
    };

    // Check if gradlew exists
    if !std::path::Path::new(&gradlew).exists() {
        return Err(format!("gradlew not found at: {}", gradlew));
    }

    let mut cmd = tokio::process::Command::new(&gradlew);
    cmd.arg(task);
    cmd.current_dir(project_path);

    // Set JAVA_HOME if provided
    if !java_home.is_empty() {
        cmd.env("JAVA_HOME", java_home);
    }

    // Add extra arguments
    if let Some(args) = extra_args {
        for arg in args {
            cmd.arg(arg);
        }
    }

    cmd.stdout(std::process::Stdio::piped());
    cmd.stderr(std::process::Stdio::piped());

    let mut child = cmd.spawn()
        .map_err(|e| format!("Failed to start Gradle: {}", e))?;

    let build_id_clone = build_id.clone();
    let app_clone = app.clone();

    // Stream stdout
    if let Some(stdout) = child.stdout.take() {
        let bid = build_id_clone.clone();
        let app_out = app_clone.clone();
        let tx = event_tx.clone();
        tokio::spawn(async move {
            let reader = BufReader::new(stdout);
            let mut lines = reader.lines();
            while let Ok(Some(line)) = lines.next_line().await {
                let event = serde_json::json!({
                    "type": "build_log",
                    "buildId": bid,
                    "line": line,
                    "stream": "stdout"
                });
                let _ = app_out.emit("build-log", event.to_string());
                let _ = tx.send(event.to_string());
            }
        });
    }

    // Stream stderr
    if let Some(stderr) = child.stderr.take() {
        let bid = build_id_clone.clone();
        let app_err = app_clone.clone();
        let tx = event_tx.clone();
        tokio::spawn(async move {
            let reader = BufReader::new(stderr);
            let mut lines = reader.lines();
            while let Ok(Some(line)) = lines.next_line().await {
                let event = serde_json::json!({
                    "type": "build_log",
                    "buildId": bid,
                    "line": line,
                    "stream": "stderr"
                });
                let _ = app_err.emit("build-log", event.to_string());
                let _ = tx.send(event.to_string());
            }
        });
    }

    // Wait for build to finish in background
    let bid_final = build_id.clone();
    let app_final = app.clone();
    let project_path_owned = project_path.to_string();
    let task_owned = task.to_string();
    let tx_final = event_tx.clone();

    tokio::spawn(async move {
        let status = child.wait().await;
        let (phase, error) = match status {
            Ok(s) if s.success() => ("complete".to_string(), None),
            Ok(s) => ("failed".to_string(), Some(format!("Gradle exited with code {:?}", s.code()))),
            Err(e) => ("failed".to_string(), Some(format!("Process error: {}", e))),
        };

        // Try to find the output APK
        let output_apk = find_output_apk(&project_path_owned, &task_owned);

        let event = serde_json::json!({
            "type": "build_status",
            "buildId": bid_final,
            "phase": phase,
            "progress": 100,
            "outputApk": output_apk,
            "error": error,
        });
        let _ = app_final.emit("build-status", event.to_string());
        let _ = tx_final.send(event.to_string());
    });

    Ok(BuildInfo {
        id: build_id,
        project_path: project_path.to_string(),
        task: task.to_string(),
        phase: "compiling".into(),
        progress: 0,
        started_at,
        finished_at: None,
        output_apk: None,
        logs: Vec::new(),
        error: None,
    })
}

/// Public wrapper for companion server to call find_output_apk.
pub fn find_output_apk_public(project_path: &str, task: &str) -> Option<String> {
    find_output_apk(project_path, task)
}

/// Try to locate the output APK after a build.
fn find_output_apk(project_path: &str, task: &str) -> Option<String> {
    let variant = if task.contains("Release") || task.contains("release") {
        "release"
    } else {
        "debug"
    };

    // Standard Gradle APK output location
    let apk_dir = format!("{}\\app\\build\\outputs\\apk\\{}", project_path, variant);
    if let Some(found) = find_file_with_ext(&apk_dir, "apk") {
        return Some(found);
    }

    // Flutter APK output location (flutter build apk goes here)
    let flutter_apk_dir = format!("{}\\build\\app\\outputs\\flutter-apk", project_path);
    let expected = format!("{}\\app-{}.apk", flutter_apk_dir, variant);
    if std::path::Path::new(&expected).exists() {
        return Some(expected);
    }
    if let Some(found) = find_file_with_ext(&flutter_apk_dir, "apk") {
        return Some(found);
    }

    // Flutter uses build/app/outputs/apk/{variant} in some versions
    let flutter_gradle_dir = format!("{}\\build\\app\\outputs\\apk\\{}", project_path, variant);
    if let Some(found) = find_file_with_ext(&flutter_gradle_dir, "apk") {
        return Some(found);
    }

    None
}

/// Find the first file with the given extension in a directory.
fn find_file_with_ext(dir: &str, ext: &str) -> Option<String> {
    if let Ok(entries) = std::fs::read_dir(dir) {
        for entry in entries.flatten() {
            let path = entry.path();
            if path.extension().map(|e| e == ext).unwrap_or(false) {
                return Some(path.to_string_lossy().to_string());
            }
        }
    }
    None
}

File: cross-platform/windows-controller/src-electron/services/gradle.ts
// Gradle build service — spawns gradlew.bat and streams output
// Ported from src-tauri/src/gradle.rs

import { spawn } from "child_process";
import { existsSync, readdirSync } from "fs";
import path from "path";
import type { BrowserWindow } from "electron";
import type { EventEmitter } from "events";
import { appState } from "./state";

// ---------------------------------------------------------------------------
// Build
// ---------------------------------------------------------------------------

export function startBuild(
  projectPath: string,
  task: string,
  javaHome: string | undefined,
  extraArgs: string | undefined,
  eventBus: EventEmitter,
  mainWindow: BrowserWindow | null,
  buildId: string
): Promise<{ outputApk: string | null }> {
  return new Promise((resolve, reject) => {
    const gradlew = path.join(projectPath, "gradlew.bat");
    if (!existsSync(gradlew)) {
      reject(new Error(`gradlew.bat not found in ${projectPath}`));
      return;
    }

    const args = [task];
    if (extraArgs) {
      args.push(...extraArgs.split(/\s+/).filter(Boolean));
    }

    const env: Record<string, string> = { ...process.env } as Record<string, string>;
    if (javaHome) env.JAVA_HOME = javaHome;

    const child = spawn(gradlew, args, {
      cwd: projectPath,
      env,
      shell: true,
    });

    const emitLog = (line: string) => {
      eventBus.emit("build-log", { buildId, line });
      mainWindow?.webContents.send("build-log", { buildId, line });
      // Also push to the in-memory build log
      const build = appState.builds.get(buildId);
      if (build) build.logs.push(line);
    };

    child.stdout?.on("data", (data: Buffer) => {
      for (const line of data.toString().split("\n")) {
        if (line.trim()) emitLog(line);
      }
    });

    child.stderr?.on("data", (data: Buffer) => {
      for (const line of data.toString().split("\n")) {
        if (line.trim()) emitLog(`[stderr] ${line}`);
      }
    });

    child.on("close", (code) => {
      const exitCode = code ?? 1; // null means killed by signal
      if (exitCode === 0) {
        const outputApk = findOutputApk(projectPath, task);
        const payload = {
          buildId,
          phase: "complete",
          progress: 100,
          outputApk,
          error: null,
        };
        eventBus.emit("build-status", payload);
        mainWindow?.webContents.send("build-status", payload);
        resolve({ outputApk });
      } else {
        const errMsg = `Gradle build failed with exit code ${exitCode}`;
        const payload = {
          buildId,
          phase: "failed",
          progress: 100,
          outputApk: null,
          error: errMsg,
        };
        eventBus.emit("build-status", payload);
        mainWindow?.webContents.send("build-status", payload);
        reject(new Error(errMsg));
      }
    });

    child.on("error", (err) => {
      reject(err);
    });
  });
}

// ---------------------------------------------------------------------------
// Find output APK
// ---------------------------------------------------------------------------

export function findOutputApk(
  projectPath: string,
  task: string
): string | null {
  // Determine build type from task name
  const isRelease = task.toLowerCase().includes("release");
  const buildType = isRelease ? "release" : "debug";

  const searchPaths = [
    path.join(projectPath, "app", "build", "outputs", "apk", buildType),
    path.join(projectPath, "app", "build", "outputs", "apk"),
    path.join(projectPath, "build", "outputs", "apk", buildType),
    path.join(projectPath, "build", "outputs", "apk"),
    // Flutter APK paths (in case a Flutter project has gradlew too)
    path.join(projectPath, "build", "app", "outputs", "flutter-apk"),
    path.join(projectPath, "build", "app", "outputs", "apk", buildType),
  ];

  for (const dir of searchPaths) {
    const apk = findFileWithExt(dir, ".apk");
    if (apk) return apk;
  }

  return null;
}

export function findFileWithExt(dir: string, ext: string): string | null {
  try {
    if (!existsSync(dir)) return null;
    const files = readdirSync(dir);
    for (const file of files) {
      if (file.endsWith(ext)) {
        return path.join(dir, file);
      }
    }
  } catch {
    // Directory doesn't exist or not readable
  }
  return null;
}
